### PR TITLE
add NodeTimeMaintenance to maintain node time

### PR DESCRIPTION
**Pull request by cyjseagull**

### CODE REVIEW

File: libsync/NodeTimeMaintenance.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+using namespace dev::sync;
+
+NodeTimeMaintenance::NodeTimeMaintenance()
+  : m_node2TimeOffset(std::make_shared<std::map<dev::h512, int64_t>>())
+{}
+
+void NodeTimeMaintenance::tryToUpdatePeerTimeInfo(SyncStatusPacket::Ptr _peerStatus)
+{
+    auto status = std::dynamic_pointer_cast<SyncStatusPacketWithAlignedTime>(_peerStatus);
```
Comment (fisco-dev) [inline libsync/NodeTimeMaintenance.cpp:35]: Codacy Issue found: [Variable 'status' is assigned a value that is never used.](https://app.codacy.com/gh/FISCO-BCOS/FISCO-BCOS/pullRequest?prid=5697841)

File: libsync/SyncMsgEngine.h [hunk 2]
```diff
@@ -86,6 +87,19 @@ class SyncMsgEngine : public std::enable_shared_from_this<SyncMsgEngine>
...
     }
 
+    void setNodeTimeMaintenance(NodeTimeMaintenance::Ptr _nodeTimeMaintenance)
+    {
+        if (m_nodeTimeMaintenance)
+        {
+            return;
+        }
+        m_timeAlignWorker =
+            std::make_shared<dev::ThreadPool>("alignT-" + std::to_string(m_groupId), 1);
```
Comment (bxq2011hust) [inline libsync/SyncMsgEngine.h:97]:
```suggestion
            std::make_shared<dev::ThreadPool>("alignTime-" + std::to_string(m_groupId), 1);

```

File: libsync/SyncMsgEngine.h [hunk 3]
```diff
@@ -125,6 +139,9 @@ class SyncMsgEngine : public std::enable_shared_from_this<SyncMsgEngine>
     std::shared_ptr<dev::ThreadPool> m_txsWorker;
     std::shared_ptr<dev::ThreadPool> m_txsSender;
     std::shared_ptr<dev::ThreadPool> m_txsReceiver;
+    std::shared_ptr<dev::ThreadPool> m_timeAlignWorker;
```
Comment (bxq2011hust) [inline libsync/SyncMsgEngine.h:142]: so many workers, maybe we can simplify these workers in the future
add a TODO tag?

File: libsync/SyncMsgPacket.cpp [hunk 4]
```diff
@@ -100,14 +100,19 @@ void SyncStatusPacketWithAlignedTime::encode()
 {
     m_rlpStream.clear();
     // add aligned time into the fields
-    prep(m_rlpStream, StatusPacket, 4) << number << genesisHash << latestHash << alignedTime;
+    prep(m_rlpStream, StatusPacket, m_itemCount);
+    // Note: here must encode int64_t type value with append(bigint)
+    // Otherwise, when the number is large the result of decode is incorrect
```
Comment (bxq2011hust) [inline libsync/SyncMsgPacket.cpp:105]: add a ut for this

File: libsync/SyncMsgPacket.cpp [hunk 5]
```diff
@@ -100,14 +100,19 @@ void SyncStatusPacketWithAlignedTime::encode()
 {
     m_rlpStream.clear();
     // add aligned time into the fields
-    prep(m_rlpStream, StatusPacket, 4) << number << genesisHash << latestHash << alignedTime;
```
Comment (bxq2011hust) [inline libsync/SyncMsgPacket.cpp:103]: not a new type of packet?
modify the packet struct will cause compatibility issue?

File: libconsensus/ConsensusEngineBase.h [hunk 6]
```diff
@@ -183,6 +184,20 @@ class ConsensusEngineBase : public Worker, virtual public ConsensusInterface
 
     KeyPair const& keyPair() { return m_keyPair; }
 
+    int64_t getAlignedTime() override
+    {
+        if (m_nodeTimeMaintenance)
+        {
+            return m_nodeTimeMaintenance->getAlignedTime();
+        }
+        return utcTime();
```
Comment (bxq2011hust) [inline libconsensus/ConsensusEngineBase.h:193]: if m_nodeTimeMaintenance is not set, return utcTime acceptable?

File: libsync/SyncMsgEngine.h (hunk 2 above)
Comment (cyjseagull) [reply libsync/SyncMsgEngine.h:97]: done.

File: libsync/SyncMsgEngine.h (hunk 3 above)
Comment (cyjseagull) [reply libsync/SyncMsgEngine.h:142]: ok.

Comment (cyjseagull) [reply libsync/SyncMsgEngine.h:142]: TODO already added.

File: libsync/SyncMsgPacket.cpp (hunk 5 above)
Comment (cyjseagull) [reply libsync/SyncMsgPacket.cpp:103]:
```
if (g_BCOSConfig.version() >= V2_6_0)
    {
        syncMsgPacketFactory = std::make_shared<SyncMsgPacketWithAlignedTimeFactory>();
        // create NodeTimeMaintenance
        m_nodeTimeMaintenance = std::make_shared<NodeTimeMaintenance>();
        syncMaster->setNodeTimeMaintenance(m_nodeTimeMaintenance);
    }
```

only when `supported_version >= v2.6.0`, enable this kind of packet with `alignedTime`, so there will not have compatibility issue.

File: libconsensus/ConsensusEngineBase.h (hunk 6 above)
Comment (cyjseagull) [reply libconsensus/ConsensusEngineBase.h:193]:
```
 void resetCurrentTime()
    {
        int64_t parentTime =
            m_blockChain->getBlockByNumber(m_blockChain->number())->header().timestamp();
        m_sealing.block->header().setTimestamp(
            std::max(parentTime + 1, m_consensusEngine->getAlignedTime()));
    }

```

the origin logic before introduce blockchain time align is return `utcTime()`, when `supported_version < v2.6.0`, reset the timestamp of the new block as the previous logic is acceptable.

File: libsync/SyncMsgPacket.cpp (hunk 4 above)
Comment (cyjseagull) [reply libsync/SyncMsgPacket.cpp:105]: ok.

**Pull request merged**